metabase/metabase
# Show data point values on static viz

> **Warning**
> There are a few hacks I make on BE to show the actual result, and those shouldn't be merged. Once the BE is properly implemented these hacks should be removed from this PR.
> - https://github.com/metabase/metabase/pull/24813/commits/f1ce21efbf11e6d3624eab286171e92d18a01ca6
> - https://github.com/metabase/metabase/pull/24813/commits/c5d1b1be685f74c95ddc7a07ebc313be142ebf29

EPIC: #24759
Issue: #20554

## Todos
- [x] Bar
- [x] Line
- [x] Area
- [x] Stacked Area
- [x] Long values (shorten numbers like 100k, 1m)
- [x] Crowded lines
- [x] Crowded bars
- [x] Crowded areas

## How to test
- Checkout this PR and add changes from these 2 commits manually (without them static viz won't render any data point values when rendering with BE)
    - https://github.com/metabase/metabase/commit/f1ce21efbf11e6d3624eab286171e92d18a01ca6
    - https://github.com/metabase/metabase/commit/c5d1b1be685f74c95ddc7a07ebc313be142ebf29
- Set up an email e.g. using `maildev` with `docker run -p 25:25 -p 80:80 -d --rm maildev/maildev:1.1.0`
- Start adding a bunch of charts with data point value enabled to a dashboard and send a dashboard subscription.
    - The reason to test with dashboard subscription because you can instantly send an email rather than having to wait for a pulse
    ![image](https://user-images.githubusercontent.com/1937582/186085758-62dfa7a2-11bb-4e60-bd57-a11ecb7d38dd.png)

---
The following todos will be addressed in a separate PR (https://github.com/metabase/metabase/pull/24938)
- [ ] Multiseries bars
- [ ] Close to chart edges labels should be visible
- [ ] Negative bars


<img width="640" alt="Screen Shot 2022-08-17 at 1 24 09 PM" src="https://user-images.githubusercontent.com/14301985/185084090-80224720-9523-4171-b02c-e62774deb914.png">


<img width="638" alt="Screen Shot 2022-08-17 at 1 24 17 PM" src="https://user-images.githubusercontent.com/14301985/185084128-4fe21a1e-591d-44cf-ac9a-68c71ee281fd.png">

---

## Results

### Bar chart
#### Dashboard
![image](https://user-images.githubusercontent.com/1937582/184912766-bd99b3d6-2f4d-4ab3-a281-970befe9884d.png)

#### Static viz
![image](https://user-images.githubusercontent.com/1937582/184912618-3d2098ed-373d-490e-bb9b-11f2f94f9a6a.png)

### Line chart
#### Dashboard
![image](https://user-images.githubusercontent.com/1937582/184912537-b75384b4-68df-4b65-be7c-5d288ea298db.png)

#### Static viz
![image](https://user-images.githubusercontent.com/1937582/184912556-8f0f9741-4760-461a-9136-f582822be145.png)

### Area chart
#### Dashboard
![image](https://user-images.githubusercontent.com/1937582/184912716-ac4d751e-a1c4-46e1-ba35-2ee2a83f4035.png)

#### Static viz
![image](https://user-images.githubusercontent.com/1937582/184912661-77d0a24f-e80d-402f-a57b-f7e74fe75a9c.png)

### Stacked area chart
![image](https://user-images.githubusercontent.com/1937582/185419380-18b88d0c-f31d-43c5-9333-e2b9aca31ad9.png)

### Automatic compact value format
![image](https://user-images.githubusercontent.com/1937582/185419575-c2045ad3-63f8-4e6d-9611-c9605ec1f087.png)

### Crowded lines
![image](https://user-images.githubusercontent.com/1937582/185645074-8e7863a1-0016-4836-b4b0-6b0268f55564.png)

### Crowded bars
![image](https://user-images.githubusercontent.com/1937582/185647141-7cf1c01e-cf79-456e-a617-bd15cefb4ad3.png)

### Crowded areas
![image](https://user-images.githubusercontent.com/1937582/185647161-a8619593-ace5-4f50-8365-6030f0c1d724.png)
